File: 2023/09/part2.ts
import { getInputLines } from "../aoc";

function getDifferencesBetweenNumbers(numbers: number[]) {
    const result: number[] = [];
    for (let i = 1; i < numbers.length; ++i) {
        result.push(numbers[i] - numbers[i - 1]);
    }
    return result;
}

function allZero(numbers: number[]) {
    return numbers.every(n => n === 0);
}

function getNextElementInSequence(line: string) {
    const numbers = line.split(" ").map(Number);
    const rows = [numbers];
    while (!allZero(rows[rows.length - 1])) {
        rows.push(getDifferencesBetweenNumbers(rows[rows.length - 1]));
    }
    let total = 0;
    for (let i = rows.length - 1; i >= 0; --i) {
        const row = rows[i];
        total = row[0] - total;
    }
    return total;
}

async function main() {
    let lines = await getInputLines(2023, 9);

    let total_of_predicted_numbers = 0;
    for (const line of lines) {
        if (!line.length) {
            continue;
        }
        total_of_predicted_numbers += getNextElementInSequence(line);
    }
    console.log(total_of_predicted_numbers);
}

main();

File: 2024/05/part2.py
#!/usr/bin/env python3

from collections import defaultdict
from graphlib import TopologicalSorter
import os
import re
import sys

parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append("..")

import aoc

lines = aoc.get_input_lines(2024, 5)
# lines = aoc.get_example_lines()

rules_lines = []
update_lines = []

adding_rules = True
for line in lines:
    if not len(line):
        adding_rules = False
        continue
    if adding_rules:
        rules_lines.append(line)
    else:
        update_lines.append(line)

must_precede = defaultdict(set)

for rule_line in rules_lines:
    split_rule = rule_line.split("|")
    before = int(split_rule[0])
    after = int(split_rule[1])
    must_precede[before].add(after)

sum = 0

for update_line in update_lines:
    split_update = update_line.split(",")
    update_page_numbers = [int(s) for s in split_update]
    if (len(update_page_numbers) % 2) == 0:
        raise Exception(f"Even number items in list! {update_page_numbers}")
    correct_order = True
    print("////", update_page_numbers)

    for i, n in enumerate(update_page_numbers):
        only_after = must_precede.get(n)
        if only_after is not None:
            print(f"Got rule {n} must precede {only_after}")
            for earlier_n in update_page_numbers[:i]:
                print("   considering earlier number:", earlier_n)
                if earlier_n in only_after:
                    correct_order = False
                    break
    if not correct_order:
        # Just extract the rules that apply for this update
        node_to_predecessors = defaultdict(set)
        for n in update_page_numbers:
            only_after = must_precede.get(n)
            if only_after is not None:
                print(
                    f"   for {n}, only_after is: {only_after} and page numbers are: {update_page_numbers}"
                )
                intersection = only_after & set(update_page_numbers)
                print("intersection is", intersection)
                if len(intersection):
                    node_to_predecessors[n].update(intersection)

        print("node_to_predecessors", node_to_predecessors)
        sorter = TopologicalSorter(node_to_predecessors)
        sorted_numbers = list(sorter.static_order())

        print("sorted_numbers is:", sorted_numbers)

        middle_number = sorted_numbers[len(sorted_numbers) // 2]
        sum += middle_number


print(sum)
